{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "e93b1fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from scipy import linalg\n",
    "from collections import Counter # might be helpful\n",
    "from sklearn.preprocessing import normalize\n",
    "\n",
    "# Constant used in this exercise\n",
    "# Represents the width of the motif\n",
    "width = 8\n",
    "\n",
    "# Helper dictionary for indexing\n",
    "nt_dict = {\"A\":0, \"C\":1, \"G\":2, \"T\":3}\n",
    "\n",
    "# Extract strings of sequences from the fasta file\n",
    "with open('motif-regions.fa', 'r') as f:\n",
    "    #ignore lines with '>', remove '\\n' from end of line\n",
    "    sequences = [seq.split('\\n')[0] for seq in f.readlines() if seq[0]!='>']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "bfd948e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Fill in all the functions below according to specifications\n",
    "#Some structure is provided inside the functions to help: feel free to modify it \n",
    "#TODO statements mark where something is missing\n",
    "\n",
    "def init_p(l, w, seqs, nt_dict):\n",
    "    '''\n",
    "    This function initializes weight matrix p_0, which represents our \n",
    "    initial guess for the profile of the motif. The dimensions of p_0\n",
    "    are (4, w+1), where w is the width of the motif. (The first column\n",
    "    of p_0 represents the background probability of each nucleotide, while \n",
    "    each subsequent column represents the probability of seeing each \n",
    "    nucleotide at that position in the motif.)\n",
    "\n",
    "    The weight matrix p_0 is initialized by considering all possible motif \n",
    "    starting positions in every sequence and counting the number of times \n",
    "    each nucleotide appears in each position of a motif. The background \n",
    "    probability of each nucleotide is initialized as uniform. The weight matrix \n",
    "    p_0 is then normalized so that each column sums to 1.\n",
    "\n",
    "    Inputs\n",
    "    - l: The length of the sequence (assumes all sequences are the same length)\n",
    "    - w: The width of the motif\n",
    "    - seqs: A list of strings representing the sequences \n",
    "    - nt_dict: A dictionary mapping each nucleotide to its index\n",
    "\n",
    "    Outputs\n",
    "    - p_0: A matrix of size (4, w+1) representing the motif profile\n",
    "    '''\n",
    "    p_0 = np.zeros((4, w+1))\n",
    "\n",
    "    # set a uniform background probability for each nucleotide\n",
    "    for i in nt_dict.keys():\n",
    "        p_0[nt_dict[i]][0] = 1/len(nt_dict)\n",
    "\n",
    "    # count nucleotides in all possible motifs for each sequence\n",
    "    for sequence in seqs:\n",
    "        # i in range(l-w+1) -> all possible motif starting positions\n",
    "        for i in range(l-w+1):\n",
    "            # j in range(w) -> position within the motif\n",
    "            for j in range(w):\n",
    "                # Fill in p_0\n",
    "                if sequence[j] == \"A\" :\n",
    "                    p_0[nt_dict[\"A\"]][j + 1] += 1;\n",
    "                elif sequence[j] == \"C\" :\n",
    "                    p_0[nt_dict[\"C\"]][j + 1] += 1;\n",
    "                elif sequence[j] == \"G\" :\n",
    "                    p_0[nt_dict[\"G\"]][j + 1] += 1;\n",
    "                elif sequence[j] == \"T\" :\n",
    "                    p_0[nt_dict[\"T\"]][j + 1] += 1;\n",
    "                \n",
    "    # normalize columns to sum to 1\n",
    "    p_0 = normalize(p_0, axis = 0, norm = 'l1')\n",
    "    return p_0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "988e0a00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_locations(l, w, p, seqs, nt_dict):\n",
    "    '''\n",
    "    This function updates z, a matrix representing the expected location of\n",
    "    the motif in each sequence. \n",
    "\n",
    "    Inputs\n",
    "    - l: The length of the sequence (assumes all sequences are the same length)\n",
    "    - w: The width of the motif\n",
    "    - p: A matrix of size (4, w+1) representing the motif profile\n",
    "    - seqs: A list of strings representing the sequences \n",
    "    - nt_dict: A dictionary mapping each nucleotide to its index\n",
    "\n",
    "    Outputs\n",
    "    - z: A matrix of size(len(seqs), l-w+1) representing the expected location\n",
    "            of the motif in each sequence. Z_ij == the probability that the motif\n",
    "            starts at position j in sequence i. \n",
    "    '''\n",
    "    z = np.zeros((len(seqs), l-w+1))\n",
    "    #consider each sequence i\n",
    "    for i, sequence in enumerate(seqs):\n",
    "        #consider each possible starting position for the motif j\n",
    "        for j in range(l-w+1):\n",
    "            #a list of conditional probabilities\n",
    "            p_vals = [] \n",
    "            #consider the nucleotide at each position in the sequence\n",
    "            for position, nt in enumerate(sequence):\n",
    "                #HINT: what is the probability of seeing nt at this position,\n",
    "                #given that the motif begins at position j?\n",
    "                #HINT: multiply each probability by 10 to avoid underflow errors (see pdf)\n",
    "                \n",
    "                if position <= j + w - 1 and position >= j:\n",
    "                    prob = 10 * (p[nt_dict[nt], position - j + 1])\n",
    "                    p_vals.append(prob)\n",
    "                else:\n",
    "                    p_vals.append(10 * p[nt_dict[nt], 0])\n",
    "\n",
    "            #update Z_ij\n",
    "            z[i][j] = np.prod(p_vals) #TODO\n",
    "    \n",
    "    # Normalize z_t so that the sum of each row is equal to 1\n",
    "    z = normalize(z, axis=1, norm='l1')\n",
    "    return z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "264b8632",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_locations_E_or_M():\n",
    "    '''\n",
    "    Is update_locations the expectation or maximization step of the EM algorithm?\n",
    "\n",
    "    Have this function return 'E' for expectation or 'M' for maximization.\n",
    "    '''\n",
    "    return 'E' #or return 'M'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "5753f177",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_profile(l, w, z, seqs, nt_dict):\n",
    "    '''\n",
    "    This function updates the motif profile p, based on an updated matrix z of the\n",
    "    expected locations of the motif in each sequence.\n",
    "\n",
    "    Note: Use a pseudo-count of 1\n",
    "\n",
    "    Inputs\n",
    "    - l: The length of the sequence (assumes all sequences are the same length)\n",
    "    - w: The width of the motif\n",
    "    - z: A matrix of size(len(seqs), l-w+1) representing the expected location\n",
    "            of the motif in each sequence. Z_ij == the probability that the motif\n",
    "            starts at position j in sequence i.\n",
    "    - seqs: A list of strings representing the sequences \n",
    "    - nt_dict: A dictionary mapping each nucleotide to its index\n",
    "\n",
    "    Outputs\n",
    "    - p: A matrix of size (4, w+1) representing the motif profile\n",
    "    '''\n",
    "    p = np.zeros((4, w+1))\n",
    "    n = np.zeros((4, w+1))\n",
    "\n",
    "    #Fill in n for k > 0\n",
    "    for k in range(1, w+1):\n",
    "        for nt in nt_dict.keys():\n",
    "            sum_z = 0 #sum of relevant probabilities Z_ij\n",
    "            for i, sequence in enumerate(seqs):\n",
    "                #Add relevant probabilities Z_ij to sum_z\n",
    "                for j in range(l-w+1):\n",
    "                    if sequence[j+k-1] == nt:\n",
    "                        sum_z += z[i][j]                \n",
    "            #update n\n",
    "            n[nt_dict[nt]][k] = sum_z\n",
    "    #Fill in n for k == 0\n",
    "    for nt in nt_dict.keys():\n",
    "        count = 0\n",
    "        for i, sequence in enumerate(seqs):\n",
    "            for j in range(l):\n",
    "                if sequence[j] == nt:\n",
    "                    count += 1\n",
    "        n[nt_dict[nt]][0] = count - np.sum(n[nt_dict[nt]])\n",
    "\n",
    "    #use n to calculate p\n",
    "    #Pseudo-count = 1\n",
    "    p = (n + 1) / (np.sum(n, axis = 0) + 4)\n",
    "    p = normalize(p, axis=0, norm = 'l1')\n",
    "            \n",
    "    return p\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "80c36f52",
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_profile_E_or_M():\n",
    "    '''\n",
    "    Is update_profile the expectation or maximization step of the EM algorithm?\n",
    "\n",
    "    Have this function return 'E' for expectation or 'M' for maximization.\n",
    "    '''\n",
    "    return 'M'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "515ce892",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_EM(w, seqs, nt_dict, epsilon=2**-64):\n",
    "    '''\n",
    "    This function runs the Expectation Maximization algorithm for motif-finding.\n",
    "\n",
    "    Hint: Use the previously defined functions init_p, update_locations, and update_motif\n",
    "\n",
    "    Inputs\n",
    "    - w: the width of the motif\n",
    "    - seqs: A list of (same-length) strings representing the sequences\n",
    "    - nt_dict: A dictionary mapping each nucleotide to its index\n",
    "    - epsilon: A small value. Terminate the algorithm if the difference between p_t \n",
    "            and p_prev is smaller than epsilon for each cell in the matrix.\n",
    "\n",
    "    Outputs\n",
    "    - p_t: A matrix of size (4, w+1) representing the motif profile at the end of\n",
    "            the EM algorithm.\n",
    "    - z_t: A matrix of size(len(seqs), l-w+1) representing the expected location\n",
    "            of the motif in each sequence at the end of the EM algorithm. \n",
    "    '''\n",
    "    l = len(seqs[0]) #length of the sequence\n",
    "    no_change = False #terminate algorithm when there is no change\n",
    "\n",
    "    #initialize p\n",
    "    p_prev = init_p(l, w, seqs, nt_dict)\n",
    "    \n",
    "    while not no_change:\n",
    "        #update z_t\n",
    "        z_t = update_locations(l, w, p_prev, seqs, nt_dict)\n",
    "        \n",
    "        #update p_t\n",
    "        p_t = update_profile(l, w, z_t, seqs, nt_dict)\n",
    "\n",
    "        #stop loop if the difference between p_t and p_prev is small enough\n",
    "        if np.linalg.norm(p_t - p_prev) < epsilon:\n",
    "            no_change = True\n",
    "        \n",
    "        #update p_prev\n",
    "        p_prev = p_t\n",
    "        p_t = normalize(p_t, axis=0, norm = 'l1')\n",
    "\n",
    "    return p_t, z_t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "da624517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running EM motif finder\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "operands could not be broadcast together with shapes (4,9) (9,1) ",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/sv/38pg6r6n7cv1p540dxmhhmtw0000gn/T/ipykernel_6560/2956076723.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'__main__'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Running EM motif finder'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mp_final\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mz_final\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrun_EM\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwidth\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequences\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnt_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepsilon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.0001\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\nFinal p:'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp_final\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/sv/38pg6r6n7cv1p540dxmhhmtw0000gn/T/ipykernel_6560/3447309839.py\u001b[0m in \u001b[0;36mrun_EM\u001b[0;34m(w, seqs, nt_dict, epsilon)\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[0;31m#initialize p\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m     \u001b[0mp_prev\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minit_p\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mseqs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnt_dict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mno_change\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/sv/38pg6r6n7cv1p540dxmhhmtw0000gn/T/ipykernel_6560/417999596.py\u001b[0m in \u001b[0;36minit_p\u001b[0;34m(l, w, seqs, nt_dict)\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m     \u001b[0;31m# normalize columns to sum to 1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 52\u001b[0;31m     \u001b[0mp_0\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnormalize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp_0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnorm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'l1'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     53\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mp_0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/var/folders/sv/38pg6r6n7cv1p540dxmhhmtw0000gn/T/ipykernel_6560/484057015.py\u001b[0m in \u001b[0;36mnormalize\u001b[0;34m(z, axis, norm)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mrow_sums\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mz\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mrow_sums\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrow_sums\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mz\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mz\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0mrow_sums\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnewaxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mz\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: operands could not be broadcast together with shapes (4,9) (9,1) "
     ]
    }
   ],
   "source": [
    "# Let's try out your code on a real example! Do not modify this segment!\n",
    "if __name__ == '__main__':\n",
    "    print('Running EM motif finder')\n",
    "    p_final, z_final = run_EM(width, sequences, nt_dict, epsilon=0.0001)\n",
    "    print('\\nFinal p:')\n",
    "    print(p_final)\n",
    "    assert(p_final.shape == (4, width+1)), \"dimensions of p_final are incorrect\"\n",
    "    assert(np.all(np.sum(p_final, axis=0) == 1.)), \"columns of p_final do not sum to 1\"\n",
    "    print('\\nFinal z:')\n",
    "    print(z_final)\n",
    "    assert(z_final.shape == (len(sequences), len(sequences[0])-width+1)), \"dimensions of z_t are incorrect\"\n",
    "    assert(np.all(np.sum(z_final, axis=1) == 1.)), \"rows of z_t do not sum to 1\"\n",
    "    # Note - if the rows of z don't add up to 1, you might be encountering an underflow error\n",
    "    # try scaling the conditional probabilities you are multiplying together by a constant value to avoid this\n",
    "    print('\\nConsensus motif:')\n",
    "    nt_lookup = {nt_dict[nt]:nt for nt in 'ACGT'}\n",
    "    best_motif = \"\".join([nt_lookup[x] for x in np.argmax(p_final[1:], axis=0)])\n",
    "    print(best_motif)\n",
    "    print('\\nPositions of motifs in sequence:')\n",
    "    motif_indices = np.argmax(z_final, axis=1)\n",
    "    print(motif_indices)\n",
    "    print('\\nMotif in each sequence:')\n",
    "    print(len(sequences[0]))\n",
    "    print(np.all([len(s)==len(sequences[0]) for s in sequences]))\n",
    "    motifs = [sequences[i][pos:pos+width] for i,pos in enumerate(motif_indices)]\n",
    "    print(motifs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11184393",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
